init.php

```diff
@@ -1,121 +1,173 @@
+        add_action( 'wp_update_nav_menu_item', [ __CLASS__, 'update_option' ], 10, 3 );
+        add_action( 'delete_post', [ __CLASS__, 'remove_visibility_meta' ], 1, 3 );
+
+    }
+
+    /**
+     * Display condition field in menu admin.
+     *
+     * @param int $item_id
+     *
+     * @return void
+     *
+     * @since 0.3.8
+     */
+    public static function wp_nav_menu_item_custom_fields( int $item_id ) : void {
+
+        $value = get_post_meta( $item_id, '_menu_item_visibility', true );
+
+        ?>
+        <p class="field-visibility description description-wide">
+            <label for="edit-menu-item-visibility-<?php echo $item_id; ?>">
+                <?php printf( __( 'Visibility logic (<a href="%s">?</a>)', 'menu-items-visibility-control' ), 'https://codex.wordpress.org/Conditional_Tags' ); ?><br>
+                <input type="text" class="widefat code" id="edit-menu-item-visibility-<?php echo $item_id; ?>" name="menu-item-visibility[<?php echo esc_attr( $item_id ); ?>]" value="<?php echo esc_attr( $value ); ?>" />
+            </label>
+        </p>
+        <?php
+
+    }
+
+    /**
+     * Update menu item visibility post meta.
+     *
+     * @param int $menu_id
+     * @param int $menu_item_db_id
+     * @param array $args
+     *
+     * @return void
+     *
+     * @since 0.3.8
+     */
+    public static function update_option( int $menu_id, int $menu_item_db_id, array $args ) : void {
+
+        $menu_item_visibility = filter_input( INPUT_POST, 'menu-item-visibility' );
+
+        if ( isset( $menu_item_visibility ) ) {
+
+            $meta_value     = get_post_meta( $menu_item_db_id, '_menu_item_visibility', true );
+            $new_meta_value = stripcslashes( $_POST['menu-item-visibility'][ $menu_item_db_id ] );
+
+            if ( '' === $new_meta_value ) {
+                delete_post_meta( $menu_item_db_id, '_menu_item_visibility', $meta_value );
+            } elseif ( $meta_value !== $new_meta_value ) {
+                update_post_meta( $menu_item_db_id, '_menu_item_visibility', $new_meta_value );
+            }
+
+        }
+
+    }
+
+    /**
+     * Checks the menu items for their visibility options and
+     * removes menu items that are not visible.
+     *
+     * @param array $items
+     * @param object $menu
+     * @param array $args
+     *
+     * @return array
+     *
+     * @since 0.1
+     */
+    public static function visibility_check( array $items, object $menu, array $args ) : array {
+
+        $hidden_items = [];
+
+        foreach ( $items as $key => $item ) {
+
+            $item_parent = get_post_meta( $item->ID, '_menu_item_menu_item_parent', true );
+            $visible     = true;
+
+            if ( $logic = get_post_meta( $item->ID, '_menu_item_visibility', true ) ) {
+
+                set_error_handler( 'self::error_handler' );
+
+                try {
+                    eval( '$visible = ' . $logic . ';' );
```

Comment:
Most importantly, `eval()` security issues must be considered carefully https://stackoverflow.com/questions/951373/when-is-eval-evil-in-php/951868

We should attempt a simple allowlist option strategy

* record allowed function / method names in an option
* run input string through https://www.php.net/manual/en/function.token-get-all.php
* if anything function / method outside allowlist is encountered, skip all execution

Give it a try. I haven't used `token_get_all()` yet, so not sure what problems it might have, but seems like it should work.

Reply:
> run input string through https://www.php.net/manual/en/function.token-get-all.php

doesnt seem to be able to explode menu condition string as it does with the php example `'<?php echo; ?>'`. will look further to see if im able to nudge it with a filter, or need to look for some other function.

Reply:
Are you manually surrounding string with `<?php` tokens?